### PR TITLE
docs: document connection and mock connection

### LINKED ISSUE #8863: Document `FooConnection`, `MakeFooConnection` in generator
https://github.com/googleapis/google-cloud-cpp/blob/d405ec7691611a8a687d2bf9f69dcc21a9171455/google/cloud/accessapproval/access_approval_connection.h#L48-L49
https://github.com/googleapis/google-cloud-cpp/blob/d405ec7691611a8a687d2bf9f69dcc21a9171455/google/cloud/accessapproval/access_approval_connection.h#L88-L90
https://github.com/googleapis/google-cloud-cpp/blob/08f1d4a76d92d79c8a5200a3c13e4453e8ad29a4/google/cloud/accessapproval/mocks/mock_access_approval_connection.h#L29-L30

As @coryan points out, these are in the public API and thus they should be documented.

### PR DESCRIPTION
**Pull request by dbolduc**
Fixes #8863 

The grammar is not perfect. In some places we say "a `AccessApproval....`"..... I don't care.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9084)

### CODE REVIEW

File: generator/internal/connection_generator.cc [hunk 1]
```diff
@@ -107,14 +107,23 @@ Status ConnectionGenerator::GenerateHeader() {
...
-    "  virtual ~$connection_class_name$() = 0;\n");
-  // clang-format on
-
   HeaderPrint(R"""(
+/**
+ * The `$connection_class_name$` object for `$client_class_name$`.
+ *
+ * This interface defines virtual methods for each of the user-facing overload
+ * sets in `$client_class_name$`. This allows users to inject custom behavior
+ * (e.g., with a Google Mock object) in a `$client_class_name$` object for use
```
Comment (coryan) [inline generator/internal/connection_generator.cc:116]: if the grammar bothers you:

```cc
/**
...
 * (e.g., with a Google Mock object) when writing tests that use objects of type `$client_class_name$`
```

File: generator/internal/connection_generator.cc [hunk 2]
```diff
@@ -192,10 +201,32 @@ Status ConnectionGenerator::GenerateHeader() {
...
-  HeaderPrint(  // clang-format off
-    "\nstd::shared_ptr<$connection_class_name$> Make$connection_class_name$(\n"
-    "    Options options = {});\n");
-  // clang-format on
+  HeaderPrint(R"""(
+/**
+ * A factory function to construct a `$connection_class_name$` object.
+ *
+ * The returned connection object should not be used directly; instead it
+ * should be given to a `$client_class_name$` instance, and methods should be
```
Comment (coryan) [inline generator/internal/connection_generator.cc:209]: "passed as an argument to the constructor of `$client_class_name$`, and that class used instead."

File: generator/internal/mock_connection_generator.cc [hunk 3]
```diff
@@ -55,9 +55,22 @@ Status MockConnectionGenerator::GenerateHeader() {
 
   // Abstract interface Connection base class
   HeaderPrint(R"""(
+/**
+ * A class to mock `$connection_class_name$`.
+ *
+ * Application developers may want to test their code with simulated responses,
+ * including errors from a `$client_class_name$`. To do so, construct a
```
Comment (coryan) [inline generator/internal/mock_connection_generator.cc:62]: s/errors from/errors, from/ ?

File: generator/internal/connection_generator.cc (hunk 1 above)
Comment (dbolduc) [reply generator/internal/connection_generator.cc]: Clever. Done.

File: generator/internal/connection_generator.cc (hunk 2 above)
Comment (dbolduc) [reply generator/internal/connection_generator.cc]: Done.

File: generator/internal/mock_connection_generator.cc (hunk 3 above)
Comment (dbolduc) [reply generator/internal/mock_connection_generator.cc]: Done.

File: generator/internal/mock_connection_generator.cc [hunk 4]
```diff
@@ -55,9 +55,22 @@ Status MockConnectionGenerator::GenerateHeader() {
...
+ * `$client_class_name$` with an instance of this class. Then use the Google
+ * Test framework functions to program the behavior of this mock.
+ *
+ * @see [This example][bq-mock] for how to test your application with GoogleTest.
+ * While the example showcases types from the BigQuery library, the underlying
+ * principles apply for any pair of `*Client` and `*Connection`.
+ *
+ * [bq-mock]: https://googleapis.dev/cpp/google-cloud-bigquery/latest/bigquery-read-mock.html
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply generator/internal/mock_connection_generator.cc:71]: Ack. There is a deeper problem. I opened #9097 to address it. I do not want to block this PR, so I am leaving the link to `latest` as is.

**Review by coryan — APPROVED** [review]

**Pull request merged**